src/main/java/appeng/helpers/InvalidPatternHelper.java

```diff
@@ -0,0 +1,157 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Applied Energistics 2.  If not, see <http://www.gnu.org/licenses/lgpl>.
+ */
+
+package appeng.helpers;
+
+
+import java.util.ArrayList;
+import java.util.List;
+
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTTagList;
+import net.minecraft.util.text.TextFormatting;
+
+import appeng.util.Platform;
+
+
+public class InvalidPatternHelper
+{
+
+	private final List<PatternIngredient> outputs = new ArrayList<>();
+	private final List<PatternIngredient> inputs = new ArrayList<>();
+	private final boolean isCrafting;
+	private final boolean canSubstitute;
+
+	public InvalidPatternHelper( final ItemStack is )
+	{
+		final NBTTagCompound encodedValue = is.getTagCompound();
+
+		if( encodedValue == null )
+		{
+			throw new IllegalArgumentException( "No pattern here!" );
+		}
+
+		final NBTTagList inTag = encodedValue.getTagList( "in", 10 );
+		final NBTTagList outTag = encodedValue.getTagList( "out", 10 );
+		this.isCrafting = encodedValue.getBoolean( "crafting" );
+
+		this.canSubstitute = this.isCrafting && encodedValue.getBoolean( "substitute" );
+
+		for( int i = 0; i < outTag.tagCount(); i++ )
+		{
+			outputs.add( new PatternIngredient( outTag.getCompoundTagAt( i ) ) );
+		}
+
+		for( int i = 0; i < inTag.tagCount(); i++ )
+		{
+			NBTTagCompound in = inTag.getCompoundTagAt( i );
+
+			// skip empty slots in the crafting grid
+			if( in.hasNoTags() )
+			{
+				continue;
+			}
+
+			inputs.add( new PatternIngredient( in ) );
+		}
+	}
+
+	public List<PatternIngredient> getOutputs()
+	{
+		return this.outputs;
+	}
+
+	public List<PatternIngredient> getInputs()
+	{
+		return this.inputs;
+	}
+
+	public boolean isCraftable()
+	{
+		return this.isCrafting;
+	}
+
+	public boolean canSubstitute()
+	{
+		return this.canSubstitute;
+	}
+
+	public class PatternIngredient
+	{
+		private String id;
+		private int count;
+		private int damage;
+
+		private ItemStack stack;
+
+		public PatternIngredient( NBTTagCompound tag )
+		{
+			this.stack = new ItemStack( tag );
+
+			if( stack.isEmpty() )
+			{
+				this.id = tag.getString( "id" );
+				this.count = tag.getByte( "Count" );
+				this.damage = Math.max( 0, tag.getShort( "Damage" ) );
+			}
+		}
+
+		public boolean isValid()
+		{
+			return !stack.isEmpty();
+		}
+
+		public String getName()
+		{
+			return isValid() ? Platform.getItemDisplayName( stack ) : id + '@' + String.valueOf( getDamage() );
+		}
+
+		public int getDamage()
+		{
+			return isValid() ? stack.getItemDamage() : damage;
+		}
+
+		public int getCount()
+		{
+			return isValid() ? stack.getCount() : count;
+		}
+
+		public ItemStack getItem() throws IllegalArgumentException
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'java.lang.IllegalArgumentException' which is a runtime exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)